Implement expression-based recursive sequence

rdbliss opened the pull request:
#### Brief description of what is fixed or changed

Introduces a new SeqBase subclass, `RecursiveSeq`, for sequences defined recursively. Specifically, any recurrence defined through a SymPy expression that depends on a finite, fixed number of previous terms is supported. This class will make it easier to implement specialized sequences with more interesting properties, such as linear recurrence relations.

Example:

```Python
In [1]: from sympy.abc import n

In [2]: from sympy import Function

In [3]: y = Function("y")

In [4]: from sympy.series.sequences import RecursiveSeq

In [5]: fib = RecursiveSeq((0, 1), y(n), y(n - 1) + y(n - 2))

In [6]: fib[:10]
Out[6]: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [7]: fib.degree
Out[7]: 2

In [8]: fib.recurrence
Out[8]: y(n - 2) + y(n - 1)

In [9]: from sympy import cos, sin

In [10]: more_complex = RecursiveSeq((0, 1), y(n), cos(y(n - 1)) + sin(y(n - 2)))

In [11]: more_complex[:5]
Out[11]: [0, 1, cos(1), sin(1) + cos(cos(1)), cos(sin(1) + cos(cos(1))) + sin(cos(1))]
```

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* series
    * implemented expression-based recursive sequence class
<!-- END RELEASE NOTES -->

rdbliss committed:
Sequences: Implement recursive sequence subclass

Introduces a new SeqBase subclass, `RecursiveSeq`, for sequences defined
recursively. Specifically, sequences that depend on a finite, fixed
number of their previous terms are supported. Any recurrence that is
expressible as a SymPy expression should work. This class will make it
easier to implement specialized sequences with more interesting
properties, such as linear recurrence relations.

Examples:

In [1]: from sympy.abc import n

In [2]: from sympy import Function

In [3]: y = Function("y")

In [4]: from sympy.series.sequences import RecursiveSeq

In [5]: fib = RecursiveSeq([0, 1], y(n), y(n - 1) + y(n - 2))

In [6]: fib[:10]
Out[6]: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [7]: fib.degree
Out[7]: 2

In [8]: fib.recurrence
Out[8]: y(n - 2) + y(n - 1)

In [9]: from sympy import cos, sin

In [10]: more_complex = RecursiveSeq([0, 1], y(n), cos(y(n - 1)) + sin(y(n - 2)))

In [11]: more_complex[:5]
Out[11]: [0, 1, cos(1), sin(1) + cos(cos(1)), cos(sin(1) + cos(cos(1))) + sin(cos(1))]

rdbliss committed:
Add test for RecursiveSeq and make it pass

rdbliss committed:
Change __init__ to __new__ in recursive sequence

This makes the args-test actually pass for RecursiveSeq.

sidhantnagpal reviewed (commented):
This looks like a good direction for recurrences. One suggestion would be to possibly use `sympy.discrete.recurrences:linrec` for evaluation of C-finite recurrences. It is faster (better asymptotic complexity) and also supports larger points of evaluation (cache size could be the blocker in the current approach).

sidhantnagpal commented on sympy/series/sequences.py:
```diff
@@ -711,6 +712,137 @@ def coeff_mul(self, coeff):
+    inspect recurrence
+
+    >>> fib.recurrence
+    y(n - 2) + y(n - 1)
```
Would it be better to have `Eq(y(n), y(n - 2) + y(n - 1))`?

rwbogl replied:
Do you mean the initial recurrence argument or just returning the recurrence for inspection from `seq.recurrence`?

rwbogl replied:
I think this is a good idea. I changed the output to match this.

sidhantnagpal commented on sympy/series/sequences.py:
```diff
@@ -711,6 +712,137 @@ def coeff_mul(self, coeff):
+        """Interval on which sequence is defined."""
+        return (self._start, self._stop)
+
+    def _eval_coeff(self, index):
```
Could this method be simplified using `@recurrence_memo`?

rwbogl replied:
It likely could; I'll check that decorator out more closely and see if it applies.

rwbogl replied:
I implemented a version with this, but it's a little clunky.

- To pass `@recurrence_memo` the initial values, `_eval_coeff` needs to be an inline function in `__new__`.

- Since `@recurrence_memo` indexes from 0, it's necessary to wrap the decorated function to shift indices down, and then shift them back up in the function itself.

If there was a version of `@recurrence_memo` that used a dictionary, or handled arbitrary starting indices, then this would be much cleaner.

rwbogl commented:
@sidhantnagpal 

> One suggestion would be to possibly use sympy.discrete.recurrences:linrec for evaluation of C-finite recurrences.

This was my idea after getting this base recursive class up and running. Once it's in, C-finite (and P-finite) sequences can be implemented with some special evaluations / other properties. Many of them already exist as functions in other places, like `sympy.discrete.recurrences:linrec`.

sidhantnagpal commented:
> Once it's in, C-finite (and P-finite) sequences can be implemented with some special evaluations / other properties.

I am not sure if there is one for P-finite. In that case, the current approach would be fine.

rdbliss committed:
Return equation from RecursiveSeq for recurrence

This also adds some docstrings to clarify what the argument for
`recurrence` should be.

sidhantnagpal commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+    >>> n = symbols("n")
+    >>> fib = RecursiveSeq((0, 1), y(n), y(n - 1) + y(n - 2))
+
+    For value at a particular point
```
Perhaps, this line (and similar) could be written as comments with the doctests?

sidhantnagpal commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
 
 
+class RecursiveSeq(SeqBase):
+
```
This blank line can be removed.

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+        return (self._start, self._stop)
+
+    def _eval_coeff(self, index):
+        if index < len(self.cache):
```
Should this test depend on `start`?

rwbogl replied:
Yes, it should be `index - start < len(self.cache)`, or just `y(index) in self.cache`. I'll fix this in an upcoming commit.

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+
+        seq = Basic.__new__(cls, initial, y(n), recurrence, start, stop)
+
+        seq.cache = {y(start + k): init for k, init in enumerate(initial)}
```
Could the cache be a list?

jksuom replied:
`dict` is probably more efficient as it can be an argument of `xreplace`.

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+    """
+
+    def __new__(cls, initial, y, recurrence, start=0, stop=oo):
+        n = y.args[0]
```
There should be some type checking to avoid this raising an exception with generic error message.

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+
+    """
+
+    def __new__(cls, initial, y, recurrence, start=0, stop=oo):
```
Perhaps `initial` could come after `recurrence`?

rwbogl replied:
Sounds good to me. I may also add logic to set dummy initial values if none are given, since the degree is automatically determined.

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+        if len(initial) != degree:
+            raise ValueError("Number of initial terms must equal degree")
+
+        degree = S(degree)
```
Could this also be `sympify(degree)`? (Or maybe `Integer(degree)`?)

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +713,160 @@ def coeff_mul(self, coeff):
+        n = y.args[0]
+        y = y.func
+        k = Wild("k", exclude=(n,))
+        degree = -oo
```
Could this be `degree = 0`?

rwbogl replied:
Yes, since anything but constant, negative shifts raises an exception. Will adjust this in upcoming commit.

rdbliss committed:
Clean RecursiveSeq docstring

rdbliss committed:
Improve logic in RecursiveSeq creation

rdbliss committed:
Reorder arguments in RecursiveSeq

This also implements default dummy initial values.

rdbliss committed:
Fix index calculations in RecursiveSeq

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +714,154 @@ def coeff_mul(self, coeff):
+
+    """
+
+    def __new__(cls, recurrence, y, initial=None, start=0, stop=oo):
```
How about replacing `y` with hidden argument by explicit `y` and `n`? Perhaps it could also be possible to do without `stop`.

rwbogl replied:
I was following the example in `rsolve`, but I agree that the hidden argument is annoying. Would it be preferable to split the two up even if it conflicts with how `rsolve` works?

jksuom replied:
I have no strong preference either way. In any case, `n` should probably be checked to avoid arguments like `y(3)`.

rwbogl replied:
I'll look into which would be better, but check `n.is_symbol` either way.

Also, to the `stop` point, I'm worried about recurrences like `y(n) = y(n - 1) / (n - 6)`. This is defined only for, say, `0 <= n < 6`.

jksuom replied:
I think that it is up to the user to check the return value for infinities or undefined values.

rwbogl commented on sympy/series/sequences.py:
```diff
@@ -711,6 +714,154 @@ def coeff_mul(self, coeff):
+        return self.cache[self.y(self._start + current)]
+
+    @property
+    def gen(self):
```
Ah, I misunderstood the meaning of "generator" in the docstring of `SeqBase.gen`. (I thought this was strange.) I'll rework it in an upcoming commit.

rdbliss committed:
Remove stop argument to RecursiveSeq

It is now up to the user to determine any undefined or infinite return
values from their recurrence.

rdbliss committed:
Change RecursiveSeq.gen to RecursiveSeq.__iter__

rdbliss committed:
Make function argument explicit in RecursiveSeq

rdbliss committed:
Fix RecurisveSeq args to be instances of Basic

rdbliss committed:
RecursiveSeq: Use S.Infinity over oo

jksuom commented on sympy/series/sequences.py:
```diff
@@ -711,6 +714,168 @@ def coeff_mul(self, coeff):
+        then the expression should be :code:`f(a(n - 1), ..., a(n - d))`.
+
+    y : function with argument
+        The name of the recursively defined sequence, e.g., :code:`y(n)`.
```
This needs to be updated.

rwbogl replied:
Thanks for catching that; fixed in the latest commit.

rdbliss committed:
RecursiveSeq: Update docstring

jksuom commented:
This looks fine to me, thank you for your contribution.
